modules/codegen/src/smithy4s/codegen/internals/IR.scala

```diff
@@ -141,29 +142,51 @@ private[internals] case class StreamingField(
 )
 
 private[internals] object Field {
-
-  sealed trait Modifier {
-    def default: Option[Node] = this match {
-      case Modifier.RequiredDefaultMod(node, _) => Some(node)
-      case Modifier.DefaultMod(node, _)         => Some(node)
-      case Modifier.RequiredMod                 => None
-      case Modifier.NoModifier                  => None
+  sealed trait TypeModification
+
+  object TypeModification {
+    case object None extends TypeModification
+    case object Option extends TypeModification
+    case object Nullable extends TypeModification
+    case object OptionNullable extends TypeModification
+
+    implicit val order: Order[TypeModification] = Order.by {
+      case None           => 0
+      case Nullable       => 1
+      case Option         => 2
+      case OptionNullable => 3
     }
+  }
 
-    def none: Boolean = this match {
-      case Modifier.RequiredDefaultMod(_, _) => false
-      case Modifier.DefaultMod(_, _)         => false
-      case Modifier.RequiredMod              => false
-      case Modifier.NoModifier               => true
-    }
+  case class Default(node: Node, typedNode: Option[Fix[TypedNode]])
+  case class Modifier(
+      required: Boolean,
+      nullable: Boolean,
+      default: Option[Default]
+  ) {
+    def typeMod: TypeModification =
```

Comment:
This case class is unfortunately complex, so I tried to make sure most places using it didn't actually need to destructure it. The `typeMod` field is part of that, since a bunch of consumers of this class only care about whether the resulting type is wrapped in an `Option` (or `Nullable`, or `Option[Nullable[...]]`).

I did look into converting the underlying type representation to a `Nullable` before reaching the point where the modifier was generated - in which case in theory nothing here should need to change - but I was scared off by the fact that nothing in that part of the code seemed to be doing modifications of the underlying shape's type. I can have a second look if people think it's better there, though - I'm still not happy with how complex the modifier logic is now. 🤷‍♀️

Reply:
I'll give it a think, but worst case scenario I'm not too bothered : the internal representation is implementation detail, and even though the complexity does increase with the "nullable" requirement, it's still acceptable imho

Reply:
I tried to think of alternative ways of modelling this but what potentially made the code look nicer (using composable binary flags, for instance) or custom extractors also lose us exhaustive pat-mat, so ... I'm fine with the seemingly ugliness if you can't find a better way of modelling it.

Reply:
Yeah, that was pretty much my conclusion too 😓 . At least `typeMod` got rid of a bunch of the places that would have needed to destructure this case class otherwise.